File: formula-java/src/main/java/org/formula/parse/shuntingyard/ShuntingYard.java
package org.formula.parse.shuntingyard;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import org.formula.Resolvable;
import org.formula.ResolveException;
import org.formula.ResolvedValue;
import org.formula.context.DataContext;

public class ShuntingYard implements Resolvable {
    private final List<Object> stack;

    @Override
    public ResolvedValue resolve(DataContext context) {
        Deque<Object> localStack = new ArrayDeque<>();
        Deque<Object> stack = new ArrayDeque<>(this.stack);
        while (!stack.isEmpty()) {
            Object next = stack.removeFirst();
            if (next instanceof OperatorFunction0 func) {
                localStack.addFirst(func.execute());
            } else if (next instanceof OperatorFunction1 func) {
                ResolvedValue a = (ResolvedValue) checkedPopParameter(next, 1, localStack);
                localStack.addFirst(func.execute(a));
            } else if (next instanceof OperatorFunction2 func) {
                ResolvedValue b = (ResolvedValue) checkedPopParameter(next, 1, localStack);
                ResolvedValue a = (ResolvedValue) checkedPopParameter(next, 2, localStack);
                localStack.addFirst(func.execute(a, b));
            } else if (next instanceof OperatorFunction3 func) {
                ResolvedValue c = (ResolvedValue) checkedPopParameter(next, 1, localStack);
                ResolvedValue b = (ResolvedValue) checkedPopParameter(next, 2, localStack);
                ResolvedValue a = (ResolvedValue) checkedPopParameter(next, 3, localStack);
                localStack.addFirst(func.execute(a, b, c));
            } else if (next instanceof OperatorFunctionN func) {
                if (localStack.isEmpty()) {
                    throw new ResolveException("Missing arity count for \"" + func + "\"");
                }
                int arity = ((Arity) localStack.removeFirst()).arity();
                List<ResolvedValue> params = new ArrayList<>();
                while (arity-- > 0) {
                    params.add((ResolvedValue) checkedPopParameter(next, arity, localStack));
                }
                localStack.addFirst(func.execute(params));
            } else if (next instanceof Variable variable) {
                stack.addFirst(variable.get(context));
            } else if (next instanceof Comment comment) {
                localStack.addFirst(comment.fn().execute((ResolvedValue) localStack.removeFirst(), comment.text()));
            } else if (next instanceof ShuntingYard shuntingYard) {
                for (int i = shuntingYard.stack.size() - 1; i >= 0; i--) {
                    stack.addFirst(shuntingYard.stack.get(i));
                }
            } else {
                while (next instanceof Resolvable resolvable) {
                    next = resolvable.resolve(context);
                }
                localStack.addFirst(next);
            }
        }

        return (ResolvedValue) localStack.removeFirst();
    }

    private Object checkedPopParameter(Object func, int parameterIndex, Deque<Object> stack) {
        if (stack.isEmpty()) {
            throw new ResolveException("Missing parameter #" + parameterIndex + " for \"" + func + "\"");
        }
        return stack.removeFirst();
    }

    public ShuntingYard(List<Object> stack) {
        this.stack = stack;
    }
}

File: formula-cs/Formula/ShuntingYard/ShuntingYard.cs
namespace Formula.ShuntingYard;

public class ShuntingYard : IResolvable
{
    private readonly Stack<INode> _stack;
    
    public ResolvedValue Resolve(IDataContext context) {
        var localStack = new Stack<object>();
        var stack = new Stack<object>(_stack);
        while (stack.TryPop(out var next))
        {
            switch (next)
            {
                case IOperatorFunction0 func0:
                    localStack.Push(func0.Execute());
                    break;
                case IOperatorFunction1 func1:
                {
                    var a = (ResolvedValue) CheckedPopParameter(next, 1, localStack);
                    localStack.Push(func1.Execute(a));
                    break;
                }
                case IOperatorFunction2 func2:
                {
                    var b = (ResolvedValue) CheckedPopParameter(next, 1, localStack);
                    var a = (ResolvedValue) CheckedPopParameter(next, 2, localStack);
                    localStack.Push(func2.Execute(a, b));
                    break;
                }
                case IOperatorFunction3 func3:
                {
                    var c = (ResolvedValue) CheckedPopParameter(next, 1, localStack);
                    var b = (ResolvedValue) CheckedPopParameter(next, 2, localStack);
                    var a = (ResolvedValue) CheckedPopParameter(next, 3, localStack);
                    localStack.Push(func3.Execute(a, b, c));
                    break;
                }
                case IOperatorFunctionN func when localStack.Count == 0:
                    throw new ResolveException("Missing arity count for \"" + func + "\"");
                case IOperatorFunctionN func:
                {
                    var arity = ((Arity) localStack.Pop()).Value;
                    var parameters = new List<ResolvedValue>();
                    while (arity-- > 0) {
                        parameters.Add((ResolvedValue) CheckedPopParameter(next, arity, localStack));
                    }
                    localStack.Push(func.Execute(parameters));
                    break;
                }
                case Variable variable:
                    stack.Push(variable.Get(context));
                    break;
                case Comment comment:
                    localStack.Push(comment.Function((ResolvedValue) localStack.Pop(), comment.Text));
                    break;
                case ShuntingYard shuntingYard:
                    foreach (var otherNext in shuntingYard._stack)
                    {
                        stack.Push(otherNext);
                    }
                    break;
                default:
                {
                    var resolved = next;
                    while (resolved is IResolvable resolvable)
                    {
                        resolved = resolvable.Resolve(context);
                    }
                    localStack.Push(resolved);
                    break;
                }
            }
        }

        return (ResolvedValue) localStack.Pop();
    }

    private static object CheckedPopParameter(object func, int parameterIndex, Stack<object> stack) {
        if (stack.Count == 0) {
            throw new ResolveException("Missing parameter #" + parameterIndex + " for \"" + func + "\"");
        }
        return stack.Pop();
    }

    public ShuntingYard(Stack<INode> stack) {
        _stack = stack;
    }
}

public interface INode {}

public interface INamed
{
    public string Name { get; }
}

public interface IOperatorFunction0
{
    ResolvedValue Execute();
}

public interface IOperatorFunction1
{
    ResolvedValue Execute(ResolvedValue a1);
}

public interface IOperatorFunction2
{
    ResolvedValue Execute(ResolvedValue a1, ResolvedValue a2);
}

public interface IOperatorFunction3
{
    ResolvedValue Execute(ResolvedValue a1, ResolvedValue a2, ResolvedValue a3);
}

public interface IOperatorFunctionN
{
    ResolvedValue Execute(IEnumerable<ResolvedValue> values);
}

public enum Associativity
{
    Left,
    Right
}

public interface IFunction : INode, INamed {}

public interface IOperator : INode, INamed
{
    int Precedence { get; }
    Associativity Associativity { get; }
}

public record struct Arity(int Value) : INode;
public record struct Comment(string Text, Func<ResolvedValue, string, ResolvedValue> Function) : INode;

public readonly record struct Function0(string Name, Func<ResolvedValue> Function) : IFunction, IOperatorFunction0
{
    public ResolvedValue Execute()
    {
        return Function();
    }
    
    public override string ToString()
    {
        return Name;
    }
}

public readonly record struct Function1(string Name, Func<ResolvedValue, ResolvedValue> Function) : IFunction, IOperatorFunction1
{
    public ResolvedValue Execute(ResolvedValue a1)
    {
        return Function(a1);
    }
    
    public override string ToString()
    {
        return Name;
    }
}

public readonly record struct Function2(string Name, Func<ResolvedValue, ResolvedValue, ResolvedValue> Function) : IFunction, IOperatorFunction2
{
    public ResolvedValue Execute(ResolvedValue a1, ResolvedValue a2)
    {
        return Function(a1, a2);
    }
    
    public override string ToString()
    {
        return Name;
    }
}

public readonly record struct Function3(string Name, Func<ResolvedValue, ResolvedValue, ResolvedValue, ResolvedValue> Function) : IFunction, IOperatorFunction3
{
    public ResolvedValue Execute(ResolvedValue a1, ResolvedValue a2, ResolvedValue a3)
    {
        return Function(a1, a2, a3);
    }
    
    public override string ToString()
    {
        return Name;
    }
}

public readonly record struct FunctionN(string Name, Func<IEnumerable<ResolvedValue>, ResolvedValue> Function) : IFunction, IOperatorFunctionN
{
    public ResolvedValue Execute(IEnumerable<ResolvedValue> values)
    {
        return Function(values);
    }
    
    public override string ToString()
    {
        return Name;
    }
}

public readonly record struct Operator0
    (string Name, int Precedence, Associativity Associativity, Func<ResolvedValue> Function) : IOperator,
        IOperatorFunction0
{
    public ResolvedValue Execute()
    {
        return Function();
    }
    
    public override string ToString()
    {
        return Name;
    }
}

public readonly record struct Operator1
    (string Name, int Precedence, Associativity Associativity, Func<ResolvedValue, ResolvedValue> Function) : IOperator,
        IOperatorFunction1
{
    public ResolvedValue Execute(ResolvedValue a1)
    {
        return Function(a1);
    }
    
    public override string ToString()
    {
        return Name;
    }
}

public readonly record struct Operator2
    (string Name, int Precedence, Associativity Associativity, Func<ResolvedValue, ResolvedValue, ResolvedValue> Function) : IOperator,
        IOperatorFunction2
{
    public ResolvedValue Execute(ResolvedValue a1, ResolvedValue a2)
    {
        try
        {
            return Function(a1, a2);
        }
        catch (FormatException)
        {
            return ResolvedValue.NaN;
        }
    }

    public override string ToString()
    {
        return Name;
    }
}

public readonly record struct BiOperator
    (string Name, Operator1 UnaryOperator, Operator2 BinaryOperator) : INode;

internal class ResolveException : Exception
{
    public ResolveException(string message) : base(message)
    {
        
    }
}